0001-typedb/02-dev/07-response.md

```diff
@@ -0,0 +1,161 @@
+
+## Response formats
+
+[TypeDB Studio](../../02-clients/01-studio.md) and [TypeDB Console](../../02-clients/02-console.md) process the 
+responses automatically to present the results (in GUI and CLI respectively) to the user. 
+
+For [TypeDB Drivers](../../02-clients/00-clients.md#typedb-drivers): the specific methods/calls used to interpret the 
+response depend on the TypeDB Driver used. 
+
+The following is a very basic description of objects used to interpret the results from the TypeDB query response. 
+For more information please see the [API & Drivers](08-api.md) page and documentation for the 
+[Java](../../02-clients/03-java.md), 
+[Python](../../02-clients/04-python.md), and 
+[Node.js](../../02-clients/05-nodejs.md) TypeDB Drivers respectively.
+
+### ConceptMap
+
+ConceptMap is a special object, made for mapping of variables from a query to concepts in a database. Its methods 
+provide a way to interact with the concepts. 
+
+Usually represents a single solution/answer from a stream of answers for a TypeDB query.
+
+### ConceptMapGroup
+
+It’s a container for a [ConceptMap](#conceptmap) object with an owner. 
+
+### Concept
+
+Object to represent a type or an instance of a type (data) from a database. 
+
+There are separate methods for every of the base types (entity type, attribute type and relation type) or for 
+instances of every base type (entity, attribute or relation).
+
+### Numeric
+
+Numeric object represents a numeric value.
+
+### NumericGroup
+
+NumericGroup object has not only a Numeric object but also an owner.
+
+### Future
+
+Future object represents an asynchronous query result to be able to get the value later, when query execution completes. 
+
+### Explanation
+
+Explanation is a special object returned as a response to an explain query. 
+These are used to explain [data inference](06-infer.md). To perform an [explain query](06-infer.md#explain-query) 
+use `explainables` and `explainable` objects.
+
+## Number of answers
+
+If the query type can return **multiple** results (e.g. a get query) then the result of such query type is a 
+**stream/iterator** to iterate through all the results. Even if the actual query of such type returns one result or no 
+results at all — it returns a stream/iterator with one or zero iterations respectively.
+
+Query types that can return only a single answer or a void (an empty response) are executed fully asynchronous on the 
+server. To wait for a query to finish execution, and return its result if there is one, use the `get()` method of the 
+**Future** object returned by the query.
+
+## Best practice
+
+### Asynchronous queries
+
+Invoking a TypeQL query sends it to a TypeDB server, where it will be completed in the background. Local processing 
+can take place while waiting for responses to be received. Take advantage of these asynchronous queries to mask network 
+round-trip costs and increases your throughput. 
```

Comment:
Looking at the code, this seems to be correct.
I was under the impression that we returned a stream which lazily finds results? @flyingsilverfin

Reply:
> Also, is the query executed before we try to iterate over the results?

![](https://docs.vaticle.com/docs/images/client-api/tx-stream.png)